### ISSUE TITLE
Redis server crashes after upgrading RediSearch to version >= v.20.0

### ISSUE THREAD

**Issue by vruizext** [RedisLabsModules/RediSearch#170]
When there is only 1 client writing in the index everything is fine, but with > 1 clients writing in parallel, the redis server crashes almost instantly after two batches of docs are sent to  the index in parallel. 

This doesn't happen for versions <=v.0.19.5, so far I didn't face any issue with 10-15 clients writing almost simultaneously. Hence I think it might be related to the concurrent indexing, and multi threaded tokenization introduced in v.0.20.0. 

```
=== REDIS BUG REPORT START: Cut & paste starting from here ===
4401:M 21 Sep 13:17:34.849 # ------------------------------------------------
4401:M 21 Sep 13:17:34.849 # !!! Software Failure. Press left mouse button to continue
4401:M 21 Sep 13:17:34.849 # Guru Meditation: Unknown btype in unblockClient(). #blocked.c:142
4401:M 21 Sep 13:17:34.849 # (forcing SIGSEGV in order to print the stack trace)
4401:M 21 Sep 13:17:34.849 # ------------------------------------------------
4401:M 21 Sep 13:17:34.849 # Redis 4.0.1 crashed by signal: 11
4401:M 21 Sep 13:17:34.849 # Crashed running the instuction at: 0x10653d8be
4401:M 21 Sep 13:17:34.849 # Accessing address: 0xffffffffffffffff
4401:M 21 Sep 13:17:34.849 # Failed assertion: <no assertion failed> (<no file>:0)

------ STACK TRACE ------
EIP:
0   redis-server                        0x000000010653d8be _serverPanic + 334

Backtrace:
0   redis-server                        0x000000010653f3ad logStackTrace + 109
1   redis-server                        0x000000010653f75c sigsegvHandler + 236
2   libsystem_platform.dylib            0x00007fffa8580b3a _sigtramp + 26
3   ???                                 0x0000000000000400 0x0 + 1024
4   redis-server                        0x0000000106562fb7 unblockClient + 167
5   redis-server                        0x000000010656f96e moduleHandleBlockedClients + 542
6   redis-server                        0x00000001064f4feb beforeSleep + 235
7   redis-server                        0x00000001064f018e aeMain + 30
8   redis-server                        0x00000001064fae2b main + 1691
9   libdyld.dylib                       0x00007fffa8371235 start + 1

------ INFO OUTPUT ------
# Server
redis_version:4.0.1
redis_git_sha1:3a523ac3
redis_git_dirty:0
redis_build_id:d462974acd06d830
redis_mode:standalone
os:Darwin 16.6.0 x86_64
arch_bits:64
multiplexing_api:kqueue
atomicvar_api:atomic-builtin
gcc_version:4.2.1
process_id:4401
run_id:5400e63729c1b08d2b2ab92a93bdde97474c15d9
tcp_port:6380
uptime_in_seconds:68
uptime_in_days:0
hz:10
lru_clock:12820430
executable:/projects/RediSearch/redis-server
config_file:/usr/local/etc/redis-6380.conf

# Clients
connected_clients:16
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:9

# Memory
used_memory:1698720
used_memory_human:1.62M
used_memory_rss:3088384
used_memory_rss_human:2.95M
used_memory_peak:1759680
used_memory_peak_human:1.68M
used_memory_peak_perc:96.54%
used_memory_overhead:1282098
used_memory_startup:963328
used_memory_dataset:416622
used_memory_dataset_perc:56.65%
total_system_memory:8589934592
total_system_memory_human:8.00G
used_memory_lua:37888
used_memory_lua_human:37.00K
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
mem_fragmentation_ratio:1.82
mem_allocator:libc
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:0
rdb_changes_since_last_save:1
rdb_bgsave_in_progress:0
rdb_last_save_time:1505992586
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0

# Stats
total_connections_received:18
total_commands_processed:31
instantaneous_ops_per_sec:9
total_net_input_bytes:15766
total_net_output_bytes:15132
instantaneous_input_kbps:1.47
instantaneous_output_kbps:1.63
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
evicted_keys:0
keyspace_hits:466
keyspace_misses:37
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

# Replication
role:master
connected_slaves:0
master_replid:0b5f9207662b3379beb20e191269e96fd86d9bb4
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:0.06
used_cpu_user:0.05
used_cpu_sys_children:0.00
used_cpu_user_children:0.00

# Commandstats
cmdstat_multi:calls=1,usec=1,usec_per_call=1.00
cmdstat_exec:calls=11,usec=6661,usec_per_call=605.55
cmdstat_flushdb:calls=1,usec=575,usec_per_call=575.00
cmdstat_info:calls=1,usec=1298,usec_per_call=1298.00
cmdstat_monitor:calls=1,usec=5,usec_per_call=5.00

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=521,expires=0,avg_ttl=0

------ CLIENT LIST OUTPUT ------
id=454 addr=127.0.0.1:61792 fd=20 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=459 addr=127.0.0.1:61793 fd=21 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=2 addr=127.0.0.1:61765 fd=7 name= age=57 idle=0 flags=O db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=monitor
id=422 addr=127.0.0.1:61780 fd=8 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=423 addr=127.0.0.1:61781 fd=9 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=426 addr=127.0.0.1:61782 fd=10 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=427 addr=127.0.0.1:61783 fd=11 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=428 addr=127.0.0.1:61784 fd=12 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=437 addr=127.0.0.1:61785 fd=13 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=438 addr=127.0.0.1:61786 fd=14 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=441 addr=127.0.0.1:61787 fd=15 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=442 addr=127.0.0.1:61788 fd=16 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=462 addr=127.0.0.1:61794 fd=22 name= age=1 idle=0 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=32768 obl=5 oll=0 omem=0 events=r cmd=exec
id=447 addr=127.0.0.1:61789 fd=17 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=448 addr=127.0.0.1:61790 fd=18 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH
id=453 addr=127.0.0.1:61791 fd=19 name= age=1 idle=1 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=FT.SEARCH

------ REGISTERS ------
4401:M 21 Sep 13:17:34.856 #
RAX:a5450e71df0c00f0 RBX:00000001065ae673
RCX:0000200000002103 RDX:0000000000012068
RDI:00007fffb126ef78 RSI:0000000005366100
RBP:00007fff59714540 RSP:00007fff59714350
R8 :0000000000000040 R9 :00007fffb126ef70
R10:ffffffffffffffff R11:0000000000012068
R12:00007fff59714578 R13:00007fff59714578
R14:000000000000008e R15:00000001065b3fe7
RIP:000000010653d8be EFL:0000000000010206
CS :000000000000002b FS:0000000000000000  GS:0000000000000000
4401:M 21 Sep 13:17:34.856 # (00007fff5971435f) -> 000000000000000f
4401:M 21 Sep 13:17:34.856 # (00007fff5971435e) -> 00007fc675f0fa50
4401:M 21 Sep 13:17:34.856 # (00007fff5971435d) -> 00007fffa850a4de
4401:M 21 Sep 13:17:34.856 # (00007fff5971435c) -> 00007fff59714440
4401:M 21 Sep 13:17:34.856 # (00007fff5971435b) -> 0000000000000008
4401:M 21 Sep 13:17:34.856 # (00007fff5971435a) -> 00007fc675f0fa50
4401:M 21 Sep 13:17:34.856 # (00007fff59714359) -> 00007fffa850a4de
4401:M 21 Sep 13:17:34.856 # (00007fff59714358) -> 00007fff59714420
4401:M 21 Sep 13:17:34.856 # (00007fff59714357) -> 00007fffa84f3472
4401:M 21 Sep 13:17:34.856 # (00007fff59714356) -> 00007fff59714460
4401:M 21 Sep 13:17:34.856 # (00007fff59714355) -> 0000000000008003
4401:M 21 Sep 13:17:34.856 # (00007fff59714354) -> 0000000000000000
4401:M 21 Sep 13:17:34.856 # (00007fff59714353) -> 0000000000000000
4401:M 21 Sep 13:17:34.856 # (00007fff59714352) -> 0000000106627000
4401:M 21 Sep 13:17:34.856 # (00007fff59714351) -> 0000000000000002
4401:M 21 Sep 13:17:34.856 # (00007fff59714350) -> 0000000000000001

------ DUMPING CODE AROUND EIP ------
Symbol: _serverPanic (base: 0x10653d770)
Module: /usr/local/bin/redis-server (base 0x1064eb000)
$ xxd -r -p /tmp/dump.hex /tmp/dump.bin
$ objdump --adjust-vma=0x10653d770 -D -b binary -m i386:x86-64 /tmp/dump.bin
------
4401:M 21 Sep 13:17:34.856 # dump of function (hexdump of 462 bytes):
554889e541574156534881ecd80100004889d34189f64989ff84c074380f298540feffff0f298d50feffff0f299560feffff0f299d70feffff0f29a580feffff0f29ad90feffff0f29b5a0feffff0f29bdb0feffff4c898d38feffff4c898530feffff48898d28feffff488b0527d80700488b00488945e0488d8510feffff488945d0488d4510488945c8c745c430000000c745c018000000488dbdc0feffff4c8d4dc0be0001000031d2b9000100004989d8e87c390600488d1d694d080083bb5c0b000000751b488d35ee0e0700bf03040000e8675dfbffc7835c0b000001000000488d1d190e0700bf0300000031c04889dee8175ffbff488d35340e0700bf0300000031c0e8045ffbff488d355b0e0700488d95c0feffffbf0300000031c04c89f94589f0e8e45efbff488d35560e0700bf0300000031c0e8d15efbffbf0300000031c04889dee8c25efbffc60425ffffffff78488b053bd70700488b00483b45e0750e4881c4d80100005b415e415f5dc3e8973806000f1f8000000000554889e54157415641554154534881ec381000004989fe488b05fad60700488b00488945d0418b5e3883fb017534488b05e3d60700488b00483b45d00f8533110000488d35fc0007004c89f74881
Function at 0x1065a11a4 is bit_tohex
Function at 0x1064f35b0 is serverLogRaw
Function at 0x1064f3780 is serverLog
Function at 0x1065a1180 is bit_tohex

=== REDIS BUG REPORT END. Make sure to include from START to END. ===

       Please report the crash by opening an issue on github:

           http://github.com/antirez/redis/issues

  Suspect RAM error? Use redis-server --test-memory to verify it.

Segmentation fault: 11
```

**Comment by dvirsky** [comment]
Is there a timeout defined on your client connection or something? Also, have you built it with make distclean to make sure you don't have junk old stuff there? This looks like something outside of the module.

**Comment by dvirsky** [comment]
Obviously we are constantly doing extensive benchmarking with multiple clients in numerous situations, up to hundreds of concurrent clients, and have never seen this issue. 

So given that the same works for me, and the place in the stack trace, this could be 2 things IMO:

1. you didn't make distclean and there are conflicting object files. run make distclean and try again just to be sure.

2. Something in the setup is different. One suspect is timeouts, my indexing benchmarks have no timeouts. This might cause the client to become NULL in high loads or something.

**Comment by vruizext** [comment]
> you didn't make distclean and there are conflicting object files. run make distclean and try again just to be sure.

> Something in the setup is different. One suspect is timeouts, my indexing benchmarks have no timeouts. This might cause the client to become NULL in high loads or something.

Good points, I didn't think about these two things.

I've built it again after a distclean, the time out was set to 2 sec, I've changed it to 10 sec, and nothing changes.

I'm using `multi .. exec` to send batches of 50-100 documents to the index. I've observed in the redis logs that the error happens right after first `exec` happens

The ruby redis client returns this error
```
/.rvm/gems/ruby-2.4.1/gems/redis-4.0.0.rc1/lib/redis/client.rb:257:in `rescue in io': Connection lost (ECONNRESET)
```

**Comment by dvirsky** [comment]
Ok, I've never tried it with multi/exec, might be a problem with it and the
async api. Of that's the case it's a bug in redis. I'll check it. In the
mean time replace the multi with a normal pipeline and I'm pretty sure it
will work. Keep me posted if it doesn't.

**Comment by vruizext** [comment]
Confirmed, the issue happens only when I use a multi pipeline.

**Comment by dvirsky** [comment]
ok, in general AFAIK blocking commands are forbidden in MULTI, and indexing now that it is multi threaded is a blocking command, like SUBSCRIBE.  It means that the same will happen with search probably. 

But redis should not crash because of this, just return an error. There was a similar crash with Lua and it was fixed in redis 4.0.2 (released today), see https://github.com/antirez/redis/commit/314043552b3e0ce5f080d0fd2169d5c32f94a222

But looking at the fix, if that is indeed the problem, it won't help. 

@antirez what is your take on that?

**Comment by vruizext** [comment]
I've tried with  redis 4.0.2 and the issue is also happening

> ok, in general AFAIK blocking commands are forbidden in MULTI, and indexing now that it is multi threaded is a blocking command, like SUBSCRIBE. It means that the same will happen with search probably.

I didn't realize about this, then I'll have to index without using multi. Actually it's not a big deal, there is not a big difference, runtimes are very similar with and without

**Comment by dvirsky** [comment]
It shouldn't have any real effect in terms of performance, it's more a consistency and atomicity thing. In any case, the solution is probably for redis to err in that case properly, and to expose an API to the module to check if we are inside a MULTI, in which case I can disable the async stuff.

Anyway I'm leaving this open for now.

**Comment by antirez** [comment]
Hello, if this only allows in MULTI it may just be that there is no handling for this case in Redis itself. I believe at some point we decided to just give control to the module about that, exporting a flag, but probably we never went ahead with that. I don't like the approach of Redis dictating what to do. The module should chose what to do, like, if still serve the command synchronously, or if to send an error lie `-ERR this command cannot be called inside MULTI/EXEC` or like other Redis commands are doing, to just do the same thing that it's done during timeouts when we are inside a MULTI/EXEC block. I'll check briefly if there is any handling about this and report.

**Comment by antirez** [comment]
I can confirm that there is no handling about this in Redis: the `HELLO.BLOCK` command inside a MULTI/EXEC block can crash exactly like the RediSearch module. In the past it was proposed to fix this with a `RedisModule_GetClientFlags` call or something like that exporting a number of flags. I'll comment where there is already a proposal in the Redis repository or open a new issue there.

**Comment by dvirsky** [comment]
thanks Salvatore. there is a discussion on an issue about lua and blocked commands, linked from this thread. it also originated from redisearch:)

**Comment by dvirsky** [comment]
@antirez well now it's linked from this thread :) https://github.com/antirez/redis/issues/4127

**Comment by dvirsky** [comment]
I'm just wondering whether redis should just expose something like RM_GetClientFlags, or whether it should expose it like RM_ClientIsLua, RM_ClientIsMulti, etc etc.

**Comment by dvirsky** [comment]
Linking the feature issue: https://github.com/antirez/redis/issues/4328

**Comment by vruizext** [comment]
Thanks @dvirsky @antirez

**Comment by dvirsky** [comment]
I've made a PR to redis implementing the flags, once it (hopefully) gets merged I'll add checks in RediSearch to avoid crashing in this situation.  https://github.com/antirez/redis/pull/4337

**Comment by dvirsky** [comment]
The PR has been merged into redis unstable, and redis will not crash if you're using that version. I'm closing this even though with 4.0.2 this of course still happens.
